File: src/main/java/kitchenpos/menu/tobe/domain/MenuName.java
package kitchenpos.menu.tobe.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import kitchenpos.exception.IllegalNameException;

import java.util.Objects;

@Embeddable
public class MenuName {
    @Column(name = "name", nullable = false)
    private String name;

    protected MenuName() {
    }

    MenuName(String menuName) {
        validateName(menuName);
        this.name = menuName;
    }

    private static void validateName(String menuName) {
        if (Objects.isNull(menuName)) {
            throw new IllegalNameException("이름은 빈값일 수 없습니다.");
        }
    }

    public String getName() {
        return name;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MenuName menuName = (MenuName) o;
        return Objects.equals(name, menuName.name);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(name);
    }
}

File: src/main/kotlin/kitchenpos/menu/tobe/domain/MenuName.kt
package kitchenpos.menu.tobe.domain

import jakarta.persistence.Column
import jakarta.persistence.Embeddable

@Embeddable
class MenuName(
    menuNamePolicy: MenuNamePolicy,

    @Column(name = "name", nullable = false)
    val name: String
) {
    init {
        menuNamePolicy.validate(name)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as MenuName

        return name == other.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }
}
